presto-geospatial-toolkit/src/main/java/com/facebook/presto/geospatial/KDBTree.java

```diff
@@ -0,0 +1,344 @@
+            .compare(first.getYMin(), second.getYMin())
+            .result();
+
+    private static final Comparator<Rectangle> BY_Y = (first, second) -> ComparisonChain.start()
+            .compare(first.getYMin(), second.getYMin())
+            .compare(first.getXMin(), second.getXMin())
+            .result();
+
+    private static final class LeafIdAllocator
+    {
+        private int nextId;
+
+        public int next()
+        {
+            return nextId++;
+        }
+    }
+
+    public static KDBTree buildKDBTree(int maxItemsPerNode, int maxLevels, Rectangle extent, List<Rectangle> items)
+    {
+        checkArgument(maxItemsPerNode > 0, "maxItemsPerNode must be > 0");
+        checkArgument(maxLevels > 0, "maxLevels must be > 0");
+        requireNonNull(extent, "extent is null");
+        requireNonNull(items, "items is null");
+        return new KDBTree(buildKDBTreeNode(maxItemsPerNode, maxLevels, 0, extent, items, new LeafIdAllocator()));
+    }
+
+    private static Node buildKDBTreeNode(int maxItemsPerNode, int maxLevels, int level, Rectangle extent, List<Rectangle> items, LeafIdAllocator leafIdAllocator)
+    {
+        checkArgument(maxItemsPerNode > 0, "maxItemsPerNode must be > 0");
+        checkArgument(maxLevels > 0, "maxLevels must be > 0");
+        checkArgument(level >= 0, "level must be >= 0");
+        requireNonNull(extent, "extent is null");
+        requireNonNull(items, "items is null");
+
+        if (items.size() <= maxItemsPerNode || level >= maxLevels) {
+            return new Node(extent, leafIdAllocator.next());
+        }
+
+        // Split over longer side
+        boolean splitVertically = extent.getWidth() >= extent.getHeight();
+        Optional<SplitResult<Node>> splitResult = trySplit(splitVertically, maxItemsPerNode, maxLevels, level, extent, items, leafIdAllocator);
+        if (!splitResult.isPresent()) {
+            // Try spitting by the other side
+            splitResult = trySplit(!splitVertically, maxItemsPerNode, maxLevels, level, extent, items, leafIdAllocator);
+        }
+
+        if (!splitResult.isPresent()) {
+            return new Node(extent, leafIdAllocator.next());
+        }
+
+        return new Node(extent, splitResult.get().getLeft(), splitResult.get().getRight());
+    }
+
+    private static final class SplitResult<T>
+    {
+        private final T left;
+        private final T right;
+
+        private SplitResult(T left, T right)
+        {
+            this.left = requireNonNull(left, "left is null");
+            this.right = requireNonNull(right, "right is null");
+        }
+
+        public T getLeft()
+        {
+            return left;
+        }
+
+        public T getRight()
+        {
+            return right;
+        }
+    }
+
+    private static Optional<SplitResult<Node>> trySplit(boolean splitVertically, int maxItemsPerNode, int maxLevels, int level, Rectangle extent, List<Rectangle> items, LeafIdAllocator leafIdAllocator)
```

Comment:
I'm not sure I like the conditional behavior based on `splitVertically`. It might be cleaner to pass an object that knows how to get the relevant dimension out of a rectangle and split a rectangle along that dimension.

Reply:
Hmm... I can't think of a good way to do that. I'm going to try in a separate commit. Let me know what you think.